include/mlir/Analysis/Liveness.h

```diff
@@ -0,0 +1,150 @@
+//===- Liveness.h - Liveness analysis for MLIR ------------------*- C++ -*-===//
+//
+// Copyright 2019 The MLIR Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// This file contains an analysis for computing liveness information from a
+// given top-level operation. The current version of the analysis uses a
+// traditional algorithm to resolve detailed live-range information about all
+// values within the specified regions. It is also possible to query liveness
+// information on block level.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_ANALYSIS_LIVENESS_H
+#define MLIR_ANALYSIS_LIVENESS_H
+
+#include <vector>
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/DenseSet.h"
+
+namespace mlir {
+class Block;
+class Region;
+class Value;
+class Operation;
+
+class LivenessBlockInfo;
+
+/// Represents an analysis for computing liveness information from a
+/// given top-level operation. The analysis iterates over all associated
+/// regions that are attached to the given top-level operation. It
+/// computes liveness information for every value and block that are
+/// included in the mentioned regions. It relies on a fixpoint iteration
+/// to compute all live-in and live-out values of all included blocks.
+/// Sample usage:
+/// Liveness liveness(topLevelOp);
+/// auto &allInValues = liveness.getLiveIn(block);
+/// auto &allOutValues = liveness.getLiveOut(block);
+/// auto allOperationsInWhichValueIsLive = liveness.resolveLiveness(value);
+/// bool lastUse = liveness.isLastUse(value, operation);
+class Liveness {
+public:
+  using OperationListT = std::vector<Operation *>;
+  using BlockMapT = llvm::DenseMap<Block *, LivenessBlockInfo>;
+  using ValueSetT = llvm::SmallDenseSet<Value *, 16>;
+
+public:
+  /// Creates a new Liveness analysis that computes liveness
+  /// information for all associated regions.
+  Liveness(Operation *op);
+
+  /// Returns the operation this analysis was constructed from.
+  Operation *getOperation() const { return operation; }
+
+  /// Gets liveness info (if any) for the given value.
+  /// This includes all operations in which the given value is live.
+  /// Note that the operations in this list are not ordered.
+  OperationListT resolveLiveness(Value *value) const;
```

Comment:
This is a very expensive operation, I would make sure to caution about it in the doc.